fingerprint2.js

```diff
@@ -95,20 +95,44 @@
       keys = this.customEntropyFunction(keys)
       this.fontsKey(keys, function (keysWithFont) {
         that.audioKey(keysWithFont, function (newKeys) {
-          var values = []
-          that.each(newKeys.data, function (pair) {
-            var value = pair.value
-            if (value && typeof value.join === 'function') {
-              values.push(value.join(';'))
-            } else {
-              values.push(value)
-            }
+          that.enumerateDevicesKey(newKeys, function (keysWithDevices) {
+            var values = []
+            that.each(keysWithDevices.data, function (pair) {
+              var value = pair.value
+              if (value && typeof value.join === 'function') {
+                values.push(value.join(';'))
+              } else {
+                values.push(value)
+              }
+            })
+            var murmur = that.x64hash128(values.join('~~~'), 31)
+            return done(murmur, keysWithDevices.data)
           })
-          var murmur = that.x64hash128(values.join('~~~'), 31)
-          return done(murmur, newKeys.data)
         })
       })
     },
+    // https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/enumerateDevices
+    enumerateDevicesKey: function (keys, done) {
+      if (this.options.excludeEnumerateDevices || !this.isEnumerateDevicesSupported()) {
+        return done(keys)
+      }
+
+      navigator.mediaDevices.enumerateDevices()
+      .then(function (devices) {
+        var enumerateDevicesFp = []
+        devices.forEach(function (device) {
+          enumerateDevicesFp.push(device.kind + ':' + device.label + 'id=' + device.deviceId)
```

Comment:
Let's put deviceId before the label for potentially fewer collisions (label may contain `id=`). How about including groupId as well?